src/rpc/notifications.cpp

```diff
@@ -0,0 +1,381 @@
+            return NotificationType::Tx;
+        else
+            return NotificationType::Unknown;
+    }
+
+    // populates a json object with all notifications in the queue
+    // returns a range to allow removing the elements from the queue
+    // after successfull transmitting
+    queueRange_t weakDequeueNotifications(UniValue& result)
+    {
+        size_t firstElement = 0;
+        size_t elementCount = 0;
+
+        LOCK(m_cs_notification_queue);
+        for (const NotificationEntry& entry : m_queue) {
+            UniValue obj = UniValue(UniValue::VOBJ);
+            obj.pushKV("type", typeToString(entry.m_type));
+            obj.pushKV("seq", entry.m_sequence_number);
+            obj.pushKV("obj", entry.m_notification);
+            result.push_back(obj);
+            elementCount++;
+        }
+        return std::make_pair(firstElement, elementCount);
+    }
+
+    // removes notifications in the given range from the queue
+    void eraseRangeFromQueue(const queueRange_t range)
+    {
+        LOCK(m_cs_notification_queue);
+        m_queue.erase(m_queue.begin() + range.first, m_queue.begin() + range.first + range.second);
+    }
+
+    // dequeues all notifications from the queue
+    void dequeueElements(UniValue& result)
+    {
+        queueRange_t range = weakDequeueNotifications(result);
+        eraseRangeFromQueue(range);
+    }
+
+    bool elementsAvailable()
+    {
+        LOCK(m_cs_notification_queue);
+        return m_queue.size() > 0;
+    }
+
+    void registerType(NotificationType type)
+    {
+        if (type == NotificationType::Unknown)
+            return;
+
+        LOCK(m_cs_notification_queue);
+        m_registered_notification_types.insert(type);
+    }
+
+    void unregisterType(NotificationType type)
+    {
+        LOCK(m_cs_notification_queue);
+        m_registered_notification_types.erase(type);
+    }
+
+    void unregisterAllTypes()
+    {
+        LOCK(m_cs_notification_queue);
+        m_registered_notification_types.clear();
+    }
+
+    void addToQueue(NotificationEntry entry)
+    {
+        LOCK(m_cs_notification_queue);
+
+        size_t queueSize = m_queue.size();
+        if (queueSize > MAX_QUEUE_SIZE) {
+            m_queue.pop_front();
+            LogPrintf("RPC Notification limit has been reached, dropping oldest element\n");
+        }
+        m_map_sequence_numbers[entry.m_type]++;
+        entry.m_sequence_number = m_map_sequence_numbers[entry.m_type];
+        m_queue.push_back(entry);
+    }
+
+    /* checks if a certain notification type is registered */
+    bool isTypeRegistered(NotificationType type)
+    {
+        LOCK(m_cs_notification_queue);
+        return (m_registered_notification_types.find(type) != m_registered_notification_types.end());
+    }
+};
+
+class NotificationQueueManager : public CValidationInterface
+{
+public:
+    CCriticalSection m_cs_queue_manager;
+    std::map<clientUUID_t, NotificationQueue*> m_map_sequence_numbers;
```

Comment:
It'd be great if we could de-duplicate the queues here - no need to have a queue per client, just have a global queue and keep track of how far delayed all the clients are in terms of the sequence number and just clean things up to the furthest-back client.

Reply:
Also, please use unique_ptr here instead of manual management and maybe remove the queue when there are no registered types (and, I suppose, the client is caught up) instead of keeping around a null queue.